### PR TITLE
Refactoring getFileLocation to be static CMakeBuilder method

### PR DESCRIPTION
**Pull request by amitrahman1026**
Addressing #35 
Updated regex in getFileLocation as well
Tested out some scenarios, but edge case of filename containing ":" still exist, maybe I could get some help with that 😅 
image

### CODE REVIEW

**Review by lingyuncai — COMMENTED** [review]
Hi @amitrahman1026, thanks for the contribution!

The refactor part LGTM.

For the regex in `getFileLocation`, another example is as below and we could want a scan starting from the last line to get `CMakeLists.txt` from `  CMakeLists.txt:15 (find_package)`, which doesn't have the `CMake Error at ` prefix.

```
CMake Error at /usr/share/test.cmake:137 (message):
  Some message here.
Call Stack (most recent call first):
  /usr/share/test.cmake:378 (FAILURE_MESSAGE)
  cmake/another_test.cmake:71 (find_package_handle_standard_args)
  CMakeLists.txt:15 (find_package)
```

**Comment by amitrahman1026** [comment]
Updated the regex to have `(?:CMake Error at \/)` as a matching group instead, so it should work even without prefix

image

File: src/builders/cmake.ts [hunk 1]
```diff
@@ -199,6 +199,33 @@ class CMakeBuilder implements IBuilder {
...
+   * @returns FileLocation or null
+   * examples: CMake Error at /home/test/tests.cmake:146 (message):
+   * examples: CMake Error at /home/test file name with spaces/tests.cmake:146:12 (message):
+   */
+  static getFileLocation(s: string, reverse: boolean): FileLocation | null {
+    let lines = s.trimStart().split("\n");
+    if (reverse) lines = lines.reverse();
+
+    // FIXME: this doesn't support file name which contains ':', but captures white space
+    const re = /(?:CMake Error at \/)?(?<file>[^:]+):(?<line>\d+)(:(?<col>\d+))?\b/;
```
Comment (lingyuncai) [inline src/builders/cmake.ts:216]: Maybe it's better to keep the `/` in `file` if exists. 
```suggestion
    const re = /(?:CMake Error at )?(?<file>[^:]+):(?<line>\d+)(:(?<col>\d+))?\b/;
```

File: src/builders/cmake.ts [hunk 2]
```diff
@@ -199,6 +199,33 @@ class CMakeBuilder implements IBuilder {
...
+    for (const line of lines) {
+      const m = line.match(re);
```
Comment (lingyuncai) [inline src/builders/cmake.ts:218]: trimStart() to remove the whitespace as the regex would capture the leading spaces into `file`, which is not expected.
```suggestion
      const m = line.trimStart().match(re);
```

File: src/advisors/cpp_exception.ts [hunk 3]
```diff
@@ -53,7 +53,8 @@ class CppExceptionAdvisor implements IAdvisor {
         if (
           fileExt.includes(path.extname(m.file)) &&
           !m.content.trim().startsWith("//") &&
-          !m.content.trim().startsWith("/*")
+          !m.content.trim().startsWith("/*") &&
+          !m.content.trim().endsWith("*/")
```
Comment (lingyuncai) [inline src/advisors/cpp_exception.ts:57]: Please file a different PR for #34 if you'd like to fix 😄

File: src/builders/cmake.ts (hunk 2 above)
Comment (amitrahman1026) [reply src/builders/cmake.ts]: Addressed!

File: src/advisors/cpp_exception.ts (hunk 3 above)
Comment (amitrahman1026) [reply src/advisors/cpp_exception.ts]: Moved it to a new PR 😄

**Review by lingyuncai — APPROVED** [review]
LGTM, thanks @amitrahman1026

**Pull request merged**